src/Symfony/Bundle/FrameworkBundle/Form/LoginType.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Form;
+
+use Symfony\Bundle\SecurityBundle\Security\FirewallMap;
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\Extension\Core\Type\PasswordType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Form\FormError;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Security\Http\Authentication\AuthenticationUtils;
+use Symfony\Component\Translation\TranslatorInterface;
+
+class LoginType extends AbstractType
+{
+    private $requestStack;
+    private $firewallMap;
+    private $authenticationUtils;
+    private $urlGenerator;
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    public function __construct(RequestStack $requestStack, FirewallMap $firewallMap, AuthenticationUtils $authenticationUtils, UrlGeneratorInterface $urlGenerator, TranslatorInterface $translator)
+    {
+        $this->requestStack = $requestStack;
+        $this->firewallMap = $firewallMap;
+        $this->authenticationUtils = $authenticationUtils;
+        $this->urlGenerator = $urlGenerator;
+        $this->translator = $translator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $firewallConfig = $this->firewallMap->getFirewallConfig($this->requestStack->getCurrentRequest());
+        if (!$firewallConfig || !$firewallConfig->isSecurityEnabled()) {
+            throw new \LogicException('You cannot use LoginType if security is not enabled');
+        }
+
+        $firewallOptions = $firewallConfig->getOptions()['form_login'];
+
+        $builder
+            ->add($firewallOptions['username_parameter'], TextType::class, ['label' => 'label.username']) // Label should be configurable
```

Comment:
What there are more fields? How would that be customized?

Reply:
A constructor parameter maybe?

Reply:
I guess one could customized the form by extending it. It would be the only way to add more fields.
I think we cannot inject the parameters like proposed because the current firewall is determined at runtime.

Reply:
If you don't take the firewall in account, and make the `LoginType` firewall-agnostic and authenticator-agnostic, then the form can be 100% customized